devito/passes/clusters/buffering.py

```diff
@@ -303,224 +270,173 @@ def callback(self, clusters, prefix, cache=None):
+    async_degree = options['buf-async-degree']
+    callback = options['buf-callback']
+
+    # {candidate buffered Function -> [Clusters that access it]}
+    bfmap = map_buffered_functions(clusters, key)
+
+    # {buffered Function -> Buffer}
+    xds = {}
+    mapper = {}
+    for f, clusters in bfmap.items():
+        exprs = flatten(c.exprs for c in clusters)
+
+        bdims = key(f, exprs)
+
+        dims = [d for d in f.dimensions if d not in bdims]
+        if len(dims) != 1:
+            raise InvalidOperator("Unsupported multi-dimensional `buffering` "
+                                  "required by `%s`" % f)
+        dim = dims.pop()
+
+        if is_buffering(exprs):
+            # Multi-level buffering
+            # NOTE: a bit rudimentary (we could go through the exprs one by one
+            # instead), but it's much shorter this way
+            buffers = [f for f in retrieve_functions(exprs) if f.is_Array]
+            assert len(buffers) == 1, "Unexpected form of multi-level buffering"
+            buffer, = buffers
+            xd = buffer.indices[dim]
         else:
-            self.subdims_mapper = {d0: v.pop()
-                                   for d0, v in self.subdims_mapper.items()}
-
-        # Build and sanity-check the buffer IterationIntervals
-        self.itintervals_mapper = {}
-        for e in accessv.mapper:
-            for i in e.ispace.itintervals:
-                v = self.itintervals_mapper.setdefault(i.dim, i.args)
-                if v != self.itintervals_mapper[i.dim]:
-                    raise NotImplementedError(
-                        "Cannot apply `buffering` as the buffered function `%s` is "
-                        "accessed over multiple, non-compatible iteration spaces "
-                        "along the Dimension `%s`" % (function.name, i.dim))
-        # Also add IterationIntervals for initialization along `x`, should `xi` be
-        # the only written Dimension in the `x` hierarchy
-        for d0, (interval, _, _) in list(self.itintervals_mapper.items()):
-            for i in d0._defines:
-                self.itintervals_mapper.setdefault(i, (interval.relaxed, (), Forward))
+            size = infer_buffer_size(f, dim, clusters)
+
+            if async_degree is not None:
+                if async_degree < size:
+                    warning("Ignoring provided asynchronous degree as it'd be "
+                            "too small for the required buffer (provided %d, "
+                            "but need at least %d for `%s`)"
+                            % (async_degree, size, f.name))
+                else:
+                    size = async_degree
+
+            # A special CustomDimension to use in place of `dim` in the buffer
+            try:
+                xd = xds[(dim, size)]
+            except KeyError:
+                name = sregistry.make_name(prefix='db')
+                xd = xds[(dim, size)] = BufferDimension(name, 0, size-1, size, dim)
```

Comment:
Might want to memoize this to only create a new dim if the dim/size changes

Reply:
it's already like that thanks to `xds`